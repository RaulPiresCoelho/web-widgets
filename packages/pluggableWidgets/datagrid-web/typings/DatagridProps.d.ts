/**
 * This file was generated from Datagrid.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
<<<<<<< HEAD
import { ActionValue, DynamicValue, EditableValue, ListValue, ListActionValue, ListAttributeValue, ListAttributeListValue, ListExpressionValue, ListReferenceValue, ListReferenceSetValue, ListWidgetValue, SelectionSingleValue, SelectionMultiValue } from "mendix";
=======
import { ActionValue, DynamicValue, EditableValue, ListValue, ListActionValue, ListAttributeValue, ListExpressionValue, ListReferenceValue, ListReferenceSetValue, ListWidgetValue, SelectionSingleValue, SelectionMultiValue } from "mendix";
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
import { Big } from "big.js";

export type ItemSelectionMethodEnum = "checkbox" | "rowClick";

export type ItemSelectionModeEnum = "toggle" | "clear";

<<<<<<< HEAD
export type LoadingTypeEnum = "spinner" | "skeleton";

=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
export type ShowContentAsEnum = "attribute" | "dynamicText" | "customContent";

export type HidableEnum = "yes" | "hidden" | "no";

export type WidthEnum = "autoFill" | "autoFit" | "manual";

export type MinWidthEnum = "auto" | "minContent" | "manual";

export type AlignmentEnum = "left" | "center" | "right";

export interface ColumnsType {
    showContentAs: ShowContentAsEnum;
<<<<<<< HEAD
    attribute?: ListAttributeValue<string | Big | boolean | Date> | ListAttributeListValue<string | Big | boolean | Date>;
=======
    attribute?: ListAttributeValue<string | Big | boolean | Date>;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    content?: ListWidgetValue;
    dynamicText?: ListExpressionValue<string>;
    exportValue?: ListExpressionValue<string>;
    header?: DynamicValue<string>;
    tooltip?: ListExpressionValue<string>;
    filter?: ReactNode;
    filterAssociation?: ListReferenceValue | ListReferenceSetValue;
    filterAssociationOptions?: ListValue;
<<<<<<< HEAD
    fetchOptionsLazy: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    filterAssociationOptionLabel?: ListExpressionValue<string>;
    visible: DynamicValue<boolean>;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    allowEventPropagation: boolean;
    width: WidthEnum;
    minWidth: MinWidthEnum;
    minWidthLimit: number;
    size: number;
    alignment: AlignmentEnum;
    columnClass?: ListExpressionValue<string>;
    wrapText: boolean;
}

export type PaginationEnum = "buttons" | "virtualScrolling" | "loadMore";

<<<<<<< HEAD
export type ShowPagingButtonsEnum = "always" | "auto";

export type PagingPositionEnum = "bottom" | "top" | "both";

=======
export type PagingPositionEnum = "bottom" | "top" | "both";

export type ShowPagingButtonsEnum = "always" | "auto";

>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
export type ShowEmptyPlaceholderEnum = "none" | "custom";

export type OnClickTriggerEnum = "single" | "double";

export type ConfigurationStorageTypeEnum = "attribute" | "localStorage";

export interface FilterListType {
    filter: ListAttributeValue<string | Big | boolean | Date>;
}

export interface ColumnsPreviewType {
    showContentAs: ShowContentAsEnum;
    attribute: string;
    content: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    dynamicText: string;
    exportValue: string;
    header: string;
    tooltip: string;
    filter: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    filterAssociation: string;
    filterAssociationOptions: {} | { caption: string } | { type: string } | null;
<<<<<<< HEAD
    fetchOptionsLazy: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    filterAssociationOptionLabel: string;
    visible: string;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    allowEventPropagation: boolean;
    width: WidthEnum;
    minWidth: MinWidthEnum;
    minWidthLimit: number | null;
    size: number | null;
    alignment: AlignmentEnum;
    columnClass: string;
    wrapText: boolean;
}

export interface FilterListPreviewType {
    filter: string;
}

export interface DatagridContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advanced: boolean;
    datasource: ListValue;
    refreshInterval: number;
    itemSelection?: SelectionSingleValue | SelectionMultiValue;
    itemSelectionMethod: ItemSelectionMethodEnum;
    itemSelectionMode: ItemSelectionModeEnum;
    showSelectAllToggle: boolean;
<<<<<<< HEAD
    loadingType: LoadingTypeEnum;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    columns: ColumnsType[];
    columnsFilterable: boolean;
    pageSize: number;
    pagination: PaginationEnum;
<<<<<<< HEAD
    showPagingButtons: ShowPagingButtonsEnum;
    showNumberOfRows: boolean;
    pagingPosition: PagingPositionEnum;
=======
    pagingPosition: PagingPositionEnum;
    showPagingButtons: ShowPagingButtonsEnum;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    loadMoreButtonCaption?: DynamicValue<string>;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder?: ReactNode;
    rowClass?: ListExpressionValue<string>;
    onClickTrigger: OnClickTriggerEnum;
    onClick?: ListActionValue;
    onSelectionChange?: ActionValue;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationStorageType: ConfigurationStorageTypeEnum;
    configurationAttribute?: EditableValue<string>;
<<<<<<< HEAD
    storeFiltersInPersonalization: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    filterList: FilterListType[];
    filtersPlaceholder?: ReactNode;
    filterSectionTitle?: DynamicValue<string>;
    exportDialogLabel?: DynamicValue<string>;
    cancelExportLabel?: DynamicValue<string>;
    selectRowLabel?: DynamicValue<string>;
}

export interface DatagridPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
<<<<<<< HEAD
    renderMode?: "design" | "xray" | "structure";
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    advanced: boolean;
    datasource: {} | { caption: string } | { type: string } | null;
    refreshInterval: number | null;
    itemSelection: "None" | "Single" | "Multi";
    itemSelectionMethod: ItemSelectionMethodEnum;
    itemSelectionMode: ItemSelectionModeEnum;
    showSelectAllToggle: boolean;
<<<<<<< HEAD
    loadingType: LoadingTypeEnum;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    columns: ColumnsPreviewType[];
    columnsFilterable: boolean;
    pageSize: number | null;
    pagination: PaginationEnum;
<<<<<<< HEAD
    showPagingButtons: ShowPagingButtonsEnum;
    showNumberOfRows: boolean;
    pagingPosition: PagingPositionEnum;
=======
    pagingPosition: PagingPositionEnum;
    showPagingButtons: ShowPagingButtonsEnum;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    loadMoreButtonCaption: string;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    rowClass: string;
    onClickTrigger: OnClickTriggerEnum;
    onClick: {} | null;
    onSelectionChange: {} | null;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationStorageType: ConfigurationStorageTypeEnum;
    configurationAttribute: string;
<<<<<<< HEAD
    storeFiltersInPersonalization: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    onConfigurationChange: {} | null;
    filterList: FilterListPreviewType[];
    filtersPlaceholder: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    filterSectionTitle: string;
    exportDialogLabel: string;
    cancelExportLabel: string;
    selectRowLabel: string;
}
