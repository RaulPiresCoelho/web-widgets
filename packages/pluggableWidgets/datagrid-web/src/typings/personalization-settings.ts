<<<<<<< HEAD
import { FilterData } from "@mendix/widget-plugin-filtering/typings/settings";
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
import { ColumnId } from "./GridColumn";
import { SortDirection, SortRule } from "./sorting";

export interface ColumnPersonalizationSettings {
    columnId: ColumnId;
    size: number | undefined;
    hidden: boolean;
    orderWeight: number;
    sortDir: SortDirection | undefined;
    sortWeight: number | undefined;
<<<<<<< HEAD
=======
    filterSettings: unknown;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}

interface ColumnPersonalizationStorageSettings {
    columnId: ColumnId;
    size: number | undefined;
    hidden: boolean;
<<<<<<< HEAD
}

export type ColumnFilterSettings = Array<[key: ColumnId, data: FilterData]>;

export type GroupFilterSettings = Array<[key: string, data: FilterData]>;

export interface GridPersonalizationStorageSettings {
    name: string;
    schemaVersion: 2;
    settingsHash: string;
    columns: ColumnPersonalizationStorageSettings[];
    groupFilters: GroupFilterSettings;
    columnFilters: ColumnFilterSettings;
=======
    filterSettings: unknown;
}

export interface GridPersonalizationStorageSettings {
    name: string;
    schemaVersion: number;
    settingsHash: string;

    columns: ColumnPersonalizationStorageSettings[];
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    columnOrder: ColumnId[];
    sortOrder: SortRule[];
}
