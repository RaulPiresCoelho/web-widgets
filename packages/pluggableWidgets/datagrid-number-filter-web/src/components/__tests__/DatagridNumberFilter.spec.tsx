import "@testing-library/jest-dom";
<<<<<<< HEAD
import { FilterAPIv2 } from "@mendix/widget-plugin-filtering/context";
import { requirePlugin } from "@mendix/widget-plugin-external-events/plugin";
import { HeaderFiltersStore, HeaderFiltersStoreProps } from "@mendix/widget-plugin-filtering/stores/HeaderFiltersStore";
import {
    actionValue,
    dynamic,
=======
import { FilterContextValue } from "@mendix/widget-plugin-filtering";
import {
    actionValue,
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    dynamicValue,
    EditableValueBuilder,
    ListAttributeValueBuilder
} from "@mendix/widget-plugin-test-utils";
import { render, screen, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createContext, createElement } from "react";

import DatagridNumberFilter from "../../DatagridNumberFilter";
import { Big } from "big.js";
<<<<<<< HEAD
import { DatagridNumberFilterContainerProps } from "../../../typings/DatagridNumberFilterProps";
import { resetIdCounter } from "downshift";

export interface StaticInfo {
    name: string;
    filtersChannelName: string;
}

const commonProps: DatagridNumberFilterContainerProps = {
=======
import { deletePlugin, requirePlugin } from "@mendix/widget-plugin-external-events/plugin";
const commonProps = {
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
    defaultFilter: "equal" as const,
    adjustable: true,
    advanced: false,
    delay: 1000
};

<<<<<<< HEAD
const headerFilterStoreInfo: StaticInfo = {
    name: commonProps.name,
    filtersChannelName: "datagrid1"
};

jest.useFakeTimers();

beforeEach(() => {
    jest.spyOn(console, "warn").mockImplementation(() => {
        // noop
    });
    resetIdCounter();
});

afterEach(() => (console.warn as jest.Mock).mockRestore());

=======
jest.useFakeTimers();

>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
describe("Number Filter", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with single attribute", () => {
<<<<<<< HEAD
            beforeEach(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withType("Long")
                                .withFormatter(
                                    value => (value ? value.toString() : ""),
                                    (value: string) => ({ valid: true, value })
                                )
                                .withFilterable(true)
                                .build()
                        }
                    ],
                    parentChannelName: headerFilterStoreInfo.filtersChannelName
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
=======
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("Long").withFilterable(true).build()
                } as FilterContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DatagridNumberFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            it("triggers attribute and onchange action on change filter value", async () => {
                const action = actionValue();
                const attribute = new EditableValueBuilder<Big>().build();
                render(<DatagridNumberFilter {...commonProps} onChange={action} valueAttribute={attribute} />);

                const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
<<<<<<< HEAD
                await user.type(screen.getByRole("spinbutton"), "10");

                act(() => {
                    jest.runOnlyPendingTimers();
                });

                expect(action.execute).toHaveBeenCalledTimes(1);
                expect(attribute.setValue).toHaveBeenCalledWith(new Big("10"));
            });

            it("clears value when external reset all event is triggered without a default value", async () => {
                const attribute = new EditableValueBuilder<Big>().build();
                const { getByRole } = render(<DatagridNumberFilter {...commonProps} valueAttribute={attribute} />);

                // First set a value
                const input = getByRole("spinbutton");
                expect(input).toHaveValue(null);

                const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
                await user.type(input, "42");

                // Run timers for the debounced setValue
                act(() => {
                    jest.runAllTimers();
                });

                expect(attribute.setValue).toHaveBeenLastCalledWith(Big(42));

                // Trigger reset event
                const plugin = requirePlugin();
                act(() => {
                    plugin.emit("datagrid1", "reset.value", false);
                });

                expect(input).toHaveValue(null);
                expect(attribute.setValue).toHaveBeenLastCalledWith(undefined);
=======

                await user.type(screen.getByRole("spinbutton"), "10");

                jest.runOnlyPendingTimers();

                expect(action.execute).toBeCalledTimes(1);
                expect(attribute.setValue).toBeCalledWith(new Big(10));
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            describe("with defaultValue", () => {
                it("initializes with defaultValue", () => {
                    render(<DatagridNumberFilter {...commonProps} defaultValue={dynamicValue<Big>(new Big(100))} />);
                    expect(screen.getByRole("spinbutton")).toHaveValue(100);
                });
<<<<<<< HEAD

=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                it("do not sync value and defaultValue when defaultValue changes from undefined to number", () => {
                    const { rerender } = render(<DatagridNumberFilter {...commonProps} defaultValue={undefined} />);
                    expect(screen.getByRole("spinbutton")).toHaveValue(null);
                    rerender(<DatagridNumberFilter {...commonProps} defaultValue={dynamicValue<Big>(new Big(100))} />);
                    expect(screen.getByRole("spinbutton")).toHaveValue(null);
                });
<<<<<<< HEAD

=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
                it("do not sync value and defaultValue when defaultValue changes from number to undefined", async () => {
                    const { rerender } = render(
                        <DatagridNumberFilter {...commonProps} defaultValue={dynamicValue<Big>(new Big(100))} />
                    );
                    expect(screen.getByRole("spinbutton")).toHaveValue(100);
                    rerender(<DatagridNumberFilter {...commonProps} defaultValue={undefined} />);
                    await waitFor(() => {
                        expect(screen.getByRole("spinbutton")).toHaveValue(100);
                    });
                });
<<<<<<< HEAD

                it("clears value when external reset all event is triggered", async () => {
                    const attribute = new EditableValueBuilder<Big>().build();
                    const value = dynamic<Big>(Big(123));
                    const { getByRole } = render(
                        <DatagridNumberFilter {...commonProps} valueAttribute={attribute} defaultValue={value} />
                    );

                    const input = getByRole("spinbutton");
                    expect(input).toHaveValue(123);

                    const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
                    // set input empty
                    await user.clear(input);
                    await user.type(input, "42");

                    act(() => {
                        jest.runAllTimers();
                    });

                    expect(attribute.setValue).toHaveBeenLastCalledWith(Big(42));

                    // Trigger reset event
                    const plugin = requirePlugin();
                    act(() => {
                        plugin.emit("datagrid1", "reset.value", true);
                    });

                    expect(input).toHaveValue(123);
                    expect(attribute.setValue).toHaveBeenLastCalledWith(Big(123));
                });
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with multiple attributes", () => {
<<<<<<< HEAD
            beforeEach(() => {
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("Long")
                                .withFormatter(
                                    value => value,
                                    () => {
                                        // noop
                                    }
                                )
                                .withFilterable(true)
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withFormatter(
                                    value => value,
                                    () => {
                                        // noop
                                    }
                                )
                                .withFilterable(true)
                                .build()
                        }
                    ],
                    parentChannelName: headerFilterStoreInfo.filtersChannelName
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
=======
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("Long")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("Decimal")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DatagridNumberFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

<<<<<<< HEAD
            it("clears value when external reset all event is triggered without a default value", async () => {
                const attribute = new EditableValueBuilder<Big>().build();
                const { getByRole } = render(<DatagridNumberFilter {...commonProps} valueAttribute={attribute} />);

                const input = getByRole("spinbutton");
                expect(input).toHaveValue(null);

                const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
                await user.type(input, "42");

                act(() => {
                    jest.runAllTimers();
                });

                expect(attribute.setValue).toHaveBeenLastCalledWith(Big(42));

                const plugin = requirePlugin();
                act(() => {
                    plugin.emit("datagrid1", "reset.value", false);
                });

                expect(input).toHaveValue(null);
                expect(attribute.setValue).toHaveBeenLastCalledWith(undefined);
            });

            it("set value when external set value event is triggered", async () => {
                const attribute = new EditableValueBuilder<Big>().build();
                const { getByRole } = render(<DatagridNumberFilter {...commonProps} valueAttribute={attribute} />);

                const input = getByRole("spinbutton");
                expect(input).toHaveValue(null);

                const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
                await user.type(input, "42");

                act(() => {
                    jest.runAllTimers();
                });

                expect(attribute.setValue).toHaveBeenLastCalledWith(Big(42));

                const plugin = requirePlugin();
                act(() => {
                    plugin.emit("filter-test", "set.value", false, {
                        numberValue: Big(100)
                    });
                });

                expect(input).toHaveValue(100);
                expect(attribute.setValue).toHaveBeenLastCalledWith(Big(100));
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
            afterAll(() => {
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with wrong attribute's type", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        { filter: new ListAttributeValueBuilder().withType("Boolean").withFilterable(true).build() }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    singleAttribute: new ListAttributeValueBuilder().withType("Boolean").withFilterable(true).build()
                } as FilterContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            it("renders error message", () => {
                const { asFragment } = render(<DatagridNumberFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with wrong multiple attributes' types", () => {
            beforeAll(() => {
<<<<<<< HEAD
                const props: HeaderFiltersStoreProps = {
                    filterList: [
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withSortable(true)
                                .build()
                        },
                        {
                            filter: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("HashString")
                                .withFilterable(true)
                                .build()
                        }
                    ]
                };
                const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                    headerFilterStore.context
                );
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                    filterDispatcher: jest.fn(),
                    multipleAttributes: {
                        attribute1: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("String")
                            .withFilterable(true)
                            .build(),
                        attribute2: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("HashString")
                            .withFilterable(true)
                            .build()
                    }
                } as FilterContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            it("renders error message", () => {
                const { asFragment } = render(<DatagridNumberFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
<<<<<<< HEAD
                (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
=======
                (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            it("renders error message", () => {
                const { asFragment } = render(<DatagridNumberFilter {...commonProps} />);

                expect(asFragment()).toMatchSnapshot();
            });
        });
    });

    describe("with multiple instances", () => {
<<<<<<< HEAD
        beforeEach(() => {
            const props: HeaderFiltersStoreProps = {
                filterList: [
                    {
                        filter: new ListAttributeValueBuilder()
                            .withType("Long")
                            .withFormatter(
                                value => value,
                                () => {
                                    // noop
                                }
                            )
                            .withFilterable(true)
                            .build()
                    }
                ]
            };
            const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
            (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
                headerFilterStore.context
            );
=======
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder().withType("Long").withFilterable(true).build()
            } as FilterContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(<DatagridNumberFilter {...commonProps} />);
            const { asFragment: fragment2 } = render(<DatagridNumberFilter {...commonProps} />);

            expect(fragment1().querySelector("button")?.getAttribute("aria-controls")).not.toBe(
                fragment2().querySelector("button")?.getAttribute("aria-controls")
            );
        });

        afterAll(() => {
<<<<<<< HEAD
            (window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });
=======
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });

    describe("events", () => {
        let dispatch: jest.Mock;
        let parentChannelName: string;
        let ctx: FilterContextValue;
        beforeEach(() => {
            dispatch = jest.fn();
            parentChannelName = Math.random().toString(36).slice(-10);
            ctx = {
                filterDispatcher: dispatch,
                eventsChannelName: parentChannelName,
                singleAttribute: new ListAttributeValueBuilder().withType("Long").withFilterable(true).build()
            };
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext(ctx);
            deletePlugin();
        });

        it("resets value on external event", async () => {
            const plugin = requirePlugin();

            expect(dispatch).toHaveBeenCalledTimes(0);

            render(
                <DatagridNumberFilter {...commonProps} defaultValue={dynamicValue<Big>(new Big(100))} name="widget_x" />
            );

            const input = screen.getByRole("spinbutton");
            expect(dispatch).toHaveBeenCalledTimes(1);
            expect(input).toHaveValue(100);

            act(() => plugin.emit("widget_x", "reset.value"));

            expect(dispatch).toHaveBeenCalledTimes(2);
            const [{ getFilterCondition }] = dispatch.mock.lastCall;
            expect(input).toHaveValue(null);
            expect(getFilterCondition()).toEqual(undefined);
        });

        it("resets value on parent event", async () => {
            const plugin = requirePlugin();

            expect(dispatch).toHaveBeenCalledTimes(0);

            render(
                <DatagridNumberFilter {...commonProps} defaultValue={dynamicValue<Big>(new Big(100))} name="widget_x" />
            );

            const input = screen.getByRole("spinbutton");
            expect(dispatch).toHaveBeenCalledTimes(1);
            expect(input).toHaveValue(100);

            act(() => plugin.emit(parentChannelName, "reset.value"));

            expect(dispatch).toHaveBeenCalledTimes(2);
            const [{ getFilterCondition }] = dispatch.mock.lastCall;
            expect(input).toHaveValue(null);
            expect(getFilterCondition()).toEqual(undefined);
        });
    });
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
});
