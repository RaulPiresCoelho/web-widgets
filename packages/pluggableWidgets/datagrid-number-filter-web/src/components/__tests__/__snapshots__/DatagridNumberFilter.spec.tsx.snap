// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Filter with single instance with multiple attributes renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="filter-selector"
    >
      <div
        class="filter-selector-content"
      >
        <button
<<<<<<< HEAD
          aria-activedescendant=""
          aria-controls="downshift-:r7:-menu"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Equal"
          class="btn btn-default filter-selector-button button-icon equal"
          id="downshift-:r7:-toggle-button"
          role="combobox"
          tabindex="0"
        >
           
        </button>
        <ul
          aria-label="Select filter type"
          class="filter-selectors hidden"
          id="downshift-:r7:-menu"
          role="listbox"
          style="position: fixed; left: 0px; top: 0px; transform: translate(0px, 0px);"
        />
=======
          aria-controls="NumberFilter1-filter-selectors"
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-default filter-selector-button button-icon equal"
        >
           
        </button>
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
      </div>
    </div>
    <input
      class="form-control filter-input"
      type="number"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Number Filter with single instance with no context renders error message 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
<<<<<<< HEAD
    The filter widget must be placed inside the column or header of the Data grid 2.0 or inside header of the Gallery widget.
=======
    The Number filter widget must be placed inside the header of the Data grid 2.0 or Gallery widget.
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
  </div>
</DocumentFragment>
`;

exports[`Number Filter with single instance with single attribute renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="filter-selector"
    >
      <div
        class="filter-selector-content"
      >
        <button
<<<<<<< HEAD
          aria-activedescendant=""
          aria-controls="downshift-:r0:-menu"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Equal"
          class="btn btn-default filter-selector-button button-icon equal"
          id="downshift-:r0:-toggle-button"
          role="combobox"
          tabindex="0"
        >
           
        </button>
        <ul
          aria-label="Select filter type"
          class="filter-selectors hidden"
          id="downshift-:r0:-menu"
          role="listbox"
          style="position: fixed; left: 0px; top: 0px; transform: translate(0px, 0px);"
        />
=======
          aria-controls="NumberFilter1-filter-selectors"
          aria-expanded="false"
          aria-haspopup="true"
          class="btn btn-default filter-selector-button button-icon equal"
        >
           
        </button>
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
      </div>
    </div>
    <input
      class="form-control filter-input"
      type="number"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`Number Filter with single instance with wrong attribute's type renders error message 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
<<<<<<< HEAD
    Unable to get filter store. Check parent widget configuration.
=======
    The attribute type being used for Number filter is not 'Autonumber, Decimal, Integer or Long'
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
  </div>
</DocumentFragment>
`;

exports[`Number Filter with single instance with wrong multiple attributes' types renders error message 1`] = `
<DocumentFragment>
  <div
    class="alert alert-danger"
  >
<<<<<<< HEAD
    Unable to get filter store. Check parent widget configuration.
=======
    The Number filter widget can't be used with the filters options you have selected. It requires a "Autonumber, Decimal, Integer or Long" attribute to be selected.
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
  </div>
</DocumentFragment>
`;
